### ISSUE TITLE
Permission denied error when trying to write to my private directory

### ISSUE THREAD

**Issue by subsetpark** [keybase/client#1920]
I installed KBFS on my work machine using my passphrase. Then this evening I downloaded the installer on my personal machine and authenticated, again using my passphrase (I wrote down my paper key but didn't use it this time). Both machines had previous versions of `keybase` installed via Homebrew.

Here's a shell log:

``` fish
home[i]> cd /keybase/private/zdsmith                                                  zax@buddhaghosa
/:zdsmith(       3)[i]> ls                                                            zax@buddhaghosa
/:zdsmith(       3)[i]> touch foo.txt                                                 zax@buddhaghosa
touch: foo.txt: Permission denied
```

I also get a OS X permission denied error when trying to copy files to the directory using the Finder.

my log id: 6ed20b455eadf1a63cd2121c

**Comment by strib** [comment]
Hi @subsetpark.  Is your work machine still on and logged into keybase?  It needs to be alive and connected in order to rekey your folders.  (I haven't checked your logs yet, so don't know for sure if this is the issue.)

How did you provision your personal machine during the login command?  It looks like you used your private PGP key (according to https://keybase.io/zdsmith/graph).  Unfortunately, we don't encrypt your file data with the PGP key, so your work machine will need to be involved to rekey your folders.  (All you should have to do it turn it on and sign in.)

**Comment by strib** [comment]
Actually, in theory it was supposed to prompt you for your paper key as well, when you first booted KBFS on the new machine.  I'll have to look more closely at the logs.

**Comment by strib** [comment]
Ok, I looked at the logs.  I'm sure that the problem is your work device isn't up, and hasn't rekeyed for you.  However, it's a mystery why you didn't get a popup asking for your paper key (you can also rekey that way).  Did you get such a popup and just ignore it?  If so, you can probably retrigger it by restarting keybase/kbfs.

**Comment by strib** [comment]
Ah, this may have been because the code is in a weird state right now -- the service side doesn't prompt for a paper key by default, but also KBFS doesn't tell it to prompt if the rekey doesn't happen after a timeout period (like 10 minutes).  We're planning to implement the latter soon, but we didn't think it was urgent.  I guess we forgot that you could sign a second device into your keychain using only your PGP key (rather than your existing device or your paper key).

@patrickxb @maxtaco: maybe we shouldn't let people sign their second devices into the keychain with a PGP key?

Anyway, @subsetpark, I'll work on a fix for this soon, but I'm not sure it'll end up in a release for Monday.  So your best bet might be waiting to get your work machine back online next week.  Very sorry for the inconvenience!

**Comment by subsetpark** [comment]
@strib thank you for such detailed analysis! It's true that I authorized both devices with my PGP (or as keybase put it, my keybase) passphrase. Though my work computer is indeed up and logged in right now, if idle. It's of course behind the standard array of network translation. Is UPnP or port forwarding required to make rekeying work?

**Comment by subsetpark** [comment]
Just as a final data point - I am able to edit the `plan.txt` in my public folder, even on my home machine.

**Comment by strib** [comment]
Hmm, weird!  The NAT shouldn't matter if the machine is already connected to our servers.  Can you please do another `keyabse log send` and post the ID?  I'm curious to see the public dir at work.  If I can't figure it out, I might need you to do a `keybase log send` from work as well.

Had you ever accessed your public dir from work?  If you accessed it for the first time from home, then no rekeying would have been necessary.  But if you acessed it from work before you provisioned your new device, then something weird is going on with rekeying.

Also, I'm about to head out on a trip for the weekend, so I might be slow to respond.  Sorry!

**Comment by maxtaco** [comment]
It could be the client at work went offline due to an auto upgrade?

**Comment by subsetpark** [comment]
1. second id log: 69cbbcd9cd74d8296b77f31c
2. I did access my public dir from work; I created a `plan.txt`, which I then edited from my home machine.

**Comment by strib** [comment]
Ok, I did some more investigation.  Side note: I forgot that public folders don't need to be rekeyed, as there is no secret that needs to be shared between devices.  Each device can sign the folder independently of the others, as long as both are in the sigchain.  That's why your public folder works.

As for the private folder -- it appears your session authentication token between your work machine and our servers expired.  (I see errors about it in our server logs.)  So, your work machine hasn't gotten the message that it should rekey.

I think we have at least one open bug about this in our internal bug tracker (@aalness is still looking into it).  Usually it would be able to make a new token automagically, but I think there's something wrong in our code.

So, in summary:
* Bug 1: Your new device should have prompted you for a paper key after 10 minutes or so of not being able to rekey.  (Fix in code review now.)
* Bug 2: Your work device should have rekeyed, but it's token expired and was not renewed. (Under investigation.)
* Temporary fix: restart keybase/KBFS on your work machine next time you have access.

Sorry for the bugs, and thanks for the report!

**Comment by subsetpark** [comment]
Awesome @strib, thanks so much! 

As a coda - I'm now about to read [The KBFS Crypto Docs](https://keybase.io/docs/kbfs-crypto), but may I suggest you add a brief gloss on exactly what rekeying is in your [Introducing...](https://keybase.io/introducing-the-keybase-filesystem) docs? The phrase is used several times there, as well as several times in the course of this ticket, but I confess I haven't figured out what it means in the sort of detail that KBFS citizenship would seem to require.

**Comment by strib** [comment]
We are working on a more advanced doc for KBFS (though less advanced than the crypto spec) with this kind of info, but it's not available yet.  Also, the next milestone release will have better user-facing error notifications when things go wrong that would have helped here.

**Comment by subsetpark** [comment]
Thanks. One final question, just for my own edification: I understand that the write-access issue is due to my work machine—the holder of the eldest sibling key, if I understand correctly—not being prompted to rekey a newly provisioned younger sibling key. 

Is it still then a mistake that I provisioned my second sibling key with my PGP key rather than my paper key? What's the difference there?

**Comment by subsetpark** [comment]
I only ask because I just deprovisioned my home machine and reprovisioned it under a new name (the things I will go to for the pursuit of knowledge, renaming my computer!) using my paper key instead of my PGP/keybase passphrase. However, I still receive the write permission issues.

**Comment by strib** [comment]
It's not a mistake that you did that.  It might be a mistake that we let you use your PGP key, but in theory if either your work machine was online, or bug (1) above was fixed, it wouldn't be a problem.

Hrmmm.   Reprovisioning with your paper key would have worked, though the problem might be that you didn't restart KBFS in between, and we don't have hooks for KBFS to know to try rekeying once the underlying device changes.  Your new device should check for rekeys once an hour, or you can force it by restarting KBFS (though don't restart Keybase, since your paper key is unlocked there!).

**Comment by subsetpark** [comment]
After restarting kbfs... 

```fish
/k/p/zdsmith $ touch foo.txt
touch: foo.txt: Input/output error
```
!

**Comment by strib** [comment]
Ooh, another secret way you can try to force it  is `echo 1 > /keybase/private/zdsmith/.kbfs_rekey`.  This is a special file that won't otherwise show up in your directory listing.

**Comment by strib** [comment]
Oh hrm.  `keybase log send` please.

**Comment by subsetpark** [comment]
log id: 156fd25ae7926fd704c6341c

**Comment by strib** [comment]
I'm seeing "Input canceled" errors.  Is it popping up a pinentry dialogue that you closed?  Maybe try `keybase unlock`?

**Comment by subsetpark** [comment]
That did bring up a dialog, but still encountering the I/O error...

log id: 2e29662dd5750e110b19311c

**Comment by strib** [comment]
Oh crap.  So, now you're in the situation of that the big yellow box on the kbfs download page warned you about.  The last writer to your folder was the rekey request written by your previous, now revoked, device.  So now it looks like your folder was written by an untrusted key, and your device is refusing to trust it.  We are in the middle of working on infrastructure that will allow you to trust writes from a key that happened before it was revoked, but we don't have it yet. Sigh.  I'm sorry, you're hitting all of our weak spots.

We need to reset your private folder because you can access it again.  I don't have the access to do that right now, but I'll ask the person who does.  It might not happen until tomorrow though.

I'm guessing your public folder is hosed now too, since you wrote to it with your previous devices.  You will probably lose the data from that folder that you wrote from the new device, sorry.

But other folders, like `/keybase/private/strib,zdsmith` should work just fine.

**Comment by subsetpark** [comment]
Ha, of course! Serves me right. Sorry for all the bother. I hope some of this was informative for you guys.

**Comment by strib** [comment]
Ok, we wiped those folders.  Can you try writing to each of them now?

**Comment by strib** [comment]
Ah, except your paper key might have expired by now.  Sigh.  Well, write if you can.

**Comment by subsetpark** [comment]
Would my paper key have expired because you had to reset my account? I thought they were pretty permanent otherwise, and signed by the device that I *didnt* revoke.

Zach Smith
http://zdsmith.com

**Comment by strib** [comment]
Oh, I just mean it would have been re-locked by your keybase daemon after an hour, and you'd have to unlock it again (but you have no way to do that, other than ANOTHER re-provision).

**Comment by subsetpark** [comment]
It worked! 
<img width="394" alt="screenshot 2016-02-07 21 44 13" src="https://cloud.githubusercontent.com/assets/96765/12877606/2517c6f4-cde4-11e5-8fc8-75a26820421e.png">
<img width="328" alt="screenshot 2016-02-07 21 45 19" src="https://cloud.githubusercontent.com/assets/96765/12877607/251a7dea-cde4-11e5-96d5-ae9c98e42ee5.png">

**Comment by strib** [comment]
Huzzah!  Thanks for bearing with us.  I'll keep this open until we merge the fix for Bug 1.

**Comment by subsetpark** [comment]
Thanks, @strib. Here's another mystery for you (not particularly pressing of course): why can I still not remove the file I touched earlier?

```fish
/k/p/zdsmith $ rm foo.txt
rm: foo.txt: Input/output error
```

log id: `7834c2ff9e12a6cebbffbb1c`

**Comment by strib** [comment]
Ah, looks like you hit a network timeout somewhere during that remove (or I guess a deadlock in our code somewhere).  It looks like some of the RPCs made it to the server, but not all of them.  I've got very bad network connectivity where I am, so I can't tell if it's a server-side problem or not.

Shouldn't be a repeatable error once the network (or whatever other transient problem was happening) is resolved.

**Comment by strib** [comment]
Ok, I've merged a fix for Bug 1, it'll be in tomorrow's builds.  A potential fix for Bug 2 is under code review.
